janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -54,6 +56,33 @@
     private static final List<Class<? extends TraversalParent>> MULTIQUERY_COMPATIBLE_STEPS =
             Arrays.asList(BranchStep.class, OptionalStep.class, RepeatStep.class, TraversalFilterStep.class);
 
+    /**
+     * This is needed as a workaround for https://issues.apache.org/jira/browse/TINKERPOP-2568. Once that is fixed
+     * and released, this utility method can be removed.
+     * @param traversal
+     * @return
+     */
+    public static StandardJanusGraph getJanusGraph(final Traversal.Admin<?, ?> traversal) {
+        Traversal.Admin<?, ?> t = traversal;
+        do {
+            Optional<Graph> optionalGraph = t.getGraph();
+            if (!optionalGraph.isPresent()) {
+                return null;
+            }
+            Graph graph = optionalGraph.get();
+            if (graph instanceof StandardJanusGraph) {
+                return (StandardJanusGraph) graph;
+            }
+            if (graph instanceof StandardJanusGraphTx) {
+                return ((StandardJanusGraphTx) graph).getGraph();
+            }
+            if (t.getParent() instanceof EmptyStep) {
```

Comment:
Do you know if `t.getParent()` can produce `null` here? If yes, I think we need to change this if to `if (t.geyParent() == null || t.getParent() instanceof EmptyStep) ...`. Maybe save `t.getParent()` in a local variable for better readability.
If this method can't produce null here, then no need to change anything

Reply:
Parent is by default EmptyStep, see [this](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/DefaultTraversal.java#L63). By checking the usage of the setter, I don't see a way to pass a null. Thus, `t.getParent()` cannot be `null`, theoretically.

What can possibly give us more confidence is the implementation of TraversalHelper, especially [this method](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/TraversalHelper.java#L546). We can see `t.getParent()` is widely used in many places without null check.

Reply:
Thank you for the confirmation!